// nodelinter/dist/services/Traverser.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Traverser = void 0;
var typescript_1 = __importStar(require("typescript"));
var Collector_1 = require("./Collector");
var Navigator_1 = require("./Navigator");
var Traverser = /** @class */ (function () {
    function Traverser() {
    }
    Traverser.traverse = function (validator) {
        var _this = this;
        return function (context) {
            return function (sourceFile) {
                _this.sourceFile = sourceFile;
                var collectorVisitor = function (node) {
                    Collector_1.Collector.run(node);
                    return typescript_1.default.visitEachChild(node, collectorVisitor, context);
                };
                var validatorVisitor = function (node) {
                    validator.setNode(node).run();
                    return typescript_1.default.visitEachChild(node, validatorVisitor, context);
                };
                typescript_1.default.visitNode(sourceFile, collectorVisitor);
                typescript_1.default.visitNode(sourceFile, validatorVisitor); // main traversal
                validator.postTraversalChecks(sourceFile);
                return sourceFile;
            };
        };
    };
    /**
     * Extract all param descriptions.
     */
    Traverser.extract = function () {
        var _this = this;
        return function (context) {
            return function (sourceFile) {
                _this.sourceFile = sourceFile;
                var extractorVisitor = function (node) {
                    if (Navigator_1.Navigator.isAssignment(node, { key: "description" })) {
                        var description = node.getChildAt(2).getText().clean();
                        if (!description)
                            return;
                        var line = typescript_1.getLineAndCharacterOfPosition(sourceFile, node.getEnd()).line;
                        line += 1;
                        Traverser.extractedDescriptions.push({
                            description: description,
                            line: line,
                            sourceFilePath: Traverser.sourceFilePath.split("n8n").pop(),
                        });
                    }
                    return typescript_1.default.visitEachChild(node, extractorVisitor, context);
                };
                typescript_1.default.visitNode(sourceFile, extractorVisitor);
                return sourceFile;
            };
        };
    };
    Traverser.sourceFilePath = "";
    Traverser.extractedDescriptions = [];
    return Traverser;
}());
exports.Traverser = Traverser;
String.prototype.unquote = function () {
    return this.replace(/(^'|'$)/g, "");
};
String.prototype.clean = function () {
    return this.trim().unquote();
};
